### ISSUE TITLE
default annotation

### ISSUE THREAD

**Issue by codalogic** [arineng/jcrvalidator#120]
(Even lower priority?)

Similar to the situation with the minExclusive and maxExclusive issue (#119), JSON Schema (+ XSD, but not in the base of Relax NG) has a `default` property.

If this were thought useful for JCR, it could be another annotation.  For example:

    "required" : @{default false} boolean ?,
    "country" : @{default "USA"} string ?,
    "state" : @{default $closed} ( "open" | $closed ) ?
    $closed = "closed"

I hope limiting the feature to primitive types would be sufficient, especially as braces can't go in annotations.  If object default were thought to be necessary, we could do something like:

    "address" : @{default $downing-street} $address ?,
    $downing-street = { "number" : 10, "street" : "Downing Street", "town" : "London" }

**Comment by anewton1998** [comment]
I don't understand the use case. What does it mean? If the value is not there then assume a default value? Why would a validator care? I think for XSD and JSON Schema, the want to address instances where tool chains are generating code. Is that where we want to go?

**Comment by codalogic** [comment]
For the use-case...

In YANG (imagine it's actually JSON!), the `leaf` rule has an optional `mandatory` member saying whether the leaf is required or not in an instance:

    leaf name {
        type string;
        description "Name of the product.";
        mandatory true;
    }

If defined in JCR, the definition would be:

    "mandatory" : boolean ?

But the question is, what happens when `mandatory` is absent?  Should the leaf be present or not?  Doing something similar to:

    "mandatory" : @{default false} boolean ?

let's the spec reader and implementer readily know what the semantics are without having to scan through the text.

The following might also work:

    "mandatory" : boolean ?, ; default: false

I agree that a validator wouldn't benefit from it, but I think humans might.

That said, I think we can wait on this one if we think there are more important things.

**Comment by anewton1998** [comment]
Ok. As long as we are clear that is really for humans and not validators, I'm good with it. Your annotation approach is my preference.
